In [1]:
import pandas as pd
import numpy as np
import requests
import tweepy
import json
import os
import re
from IPython.display import clear_output

------
<a id="indice"></a>

# Índice

1. **[Coleta](#coleta)**
    * [Twitter Archive](#coleta:twitter-archive)
    * [Image Prediction](#coleta:image-prediction)
    * [Twitter API](#coleta:twitter-api)
    
    
2. **[Avaliação](#avaliacao)**
    * [Twitter Archive](#avaliacao:twitter-archive)
    * [Image Prediction](#avaliacao:image-prediction)
    * [Twitter API](#avaliacao:twitter-api)
    * [Notas da avaliação](#avaliacao:anotacoes)
    
    
3. **[Limpeza](#limpeza)**
    * [Definição](#limpeza:definicao)
    * [Twitter Archive](#limpeza:twitter-archive)
    * [Image Prediction](#limpeza:image-prediction)
    * [Twitter API](#limpeza:twitter-api)


4. **[Armazenamento](#armazenamento)**

------
<a id="coleta"></a>

# Coleta

<a id="coleta:twitter-archive"></a>

## Coleta: Twitter archive

In [2]:
df_twitter_arc = pd.read_csv('data/twitter-archive-enhanced.csv')
df_twitter_arc.sample(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1018  746818907684614144           6.914169e+17         4.196984e+09   
1865  675354435921575936                    NaN                  NaN   

                      timestamp  \
1018  2016-06-25 21:34:37 +0000   
1865  2015-12-11 16:40:19 +0000   

                                                 source  \
1018  <a href="http://twitter.com/download/iphone" r...   
1865  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
1018  Guys... Dog Jesus 2.0\n13/10 buoyant af https:...                  NaN   
1865  Everyone needs to watch this. 13/10 https://t....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1018                       NaN                        NaN   
1865                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1018  https://twitter.com/dog_rates/status/746818907...                13   
1865  https://twitter.com/dog_rates/status/675354435...                13   

      rating_denominator  name doggo floofer pupper puppo  
1018                  10  None  None    None   None  None  
1865                  10  None  None    None   None  None

<a id="coleta:image-prediction"></a>

## Coleta: Image prediction

In [3]:
df_prediction = None

r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')

if r.status_code is 200:
    df_prediction = pd.read_csv(pd.compat.StringIO(r.text), sep='\t')    
else:    
    print('ERROR: Image prediction request returned {status_code} status code.'.format(status_code = r.status_code))

In [4]:
df_prediction.sample(2)

tweet_id                                          jpg_url  \
863  697259378236399616  https://pbs.twimg.com/media/Ca0ps3AXEAAnp9m.jpg   
390  673363615379013632  https://pbs.twimg.com/media/CVhEoq4WcAE8pBm.jpg   

     img_num          p1   p1_conf  p1_dog       p2   p2_conf  p2_dog  \
863        1  Great_Dane  0.999223    True    boxer  0.000187    True   
390        1          ox  0.193431   False  warthog  0.123827   False   

          p3   p3_conf  p3_dog  
863  whippet  0.000151    True  
390    bison  0.111177   False

<a id="coleta:twitter-api"></a>

## Coleta: Twitter API

In [5]:
# Carregar configuração do Twitter App
with open('twitter_config.json', 'r', encoding='utf-8') as file:
    app_config = json.load(file)

In [6]:
# Atribuição da configurações a variaveis locais
api_key = app_config['api_key']
api_secret = app_config['api_secret']
access_token = app_config['access_token']
access_secret = app_config['access_secret']

In [7]:
# Autentificação ao Twitter App
auth = tweepy.OAuthHandler(api_key, api_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

Buscar dados do histórico de Tweets na API

In [8]:
json_path = 'data/tweet_json.txt'
error_log = 'data/tweet_error.log'

# Salvar Tweets localmente
if(not os.path.isfile(json_path)):
    print('Aguarde. Dado o volume de requisições, a coleta de tweets pode demorar alguns minutos.')
    tweet_count = 0
    for tweet_id in df_twitter_arc.tweet_id:
        print('{percent}%'.format(percent=int((++tweet_count/df_twitter_arc.tweet_id.size)*100)))
        try:
            status = api.get_status(tweet_id)
            with open(json_path, 'a', newline='\n') as file:
                file.write(f'{json.dumps(status._json)}\n')

        except Exception as err:
            with open(error_log, 'a', newline='\n') as log:
                log.write(f'{str(tweet_id)}: {str(err.args[0])}\n')
            print(f'{str(tweet_id)}: {str(err.args[0])}')
        
        clear_output(wait=True)
else:
    print('Dados já salvos em disco, não serão executadas novas requisições à API do Twitter.')
        

Dados já salvos em disco, não serão executadas novas requisições à API do Twitter.


In [9]:
# Criar uma lista de Dictionaries com os Tweets carregados
tweets = []

with open(json_path, 'r') as file:
    for line in file:
        try:
            tweet = json.loads(line)
            
            if(tweet.get('entities', False)):
                if(tweet['entities'].get('media', False) and tweet['entities']['media'][0].get('media_url', False)):
                    tweets.append({
                        'id': tweet['id'],
                        'created_at': tweet['created_at'],
                        'in_reply_to_status_id': tweet['in_reply_to_status_id'],
                        'in_reply_to_user_id': tweet['in_reply_to_user_id'],
                        'is_quote_status': tweet['is_quote_status'],
                        'retweet_count': tweet['retweet_count'],
                        'favorite_count': tweet['favorite_count'],
                        'media_url': tweet['entities']['media'][0]['media_url'],
                        'retweeted': tweet['retweeted'],
                        'favorited': tweet['favorited']
                    })
            
            
        except Exception as e:
            print(e) 
            
        

In [10]:
# Criar dataframe de Tweets consultados na API
columns = tweets[0].keys()
df_tweets_api = pd.DataFrame(tweets, columns = columns)

------
<a id="avaliacao"></a>

# Avaliação

<a id="avaliacao:twitter-archive"></a>

## Avaliação: Twitter data archive

In [11]:
 df_twitter_arc.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
248   845397057150107648                    NaN                  NaN   
849   766008592277377025                    NaN                  NaN   
1926  674053186244734976                    NaN                  NaN   
464   817415592588222464                    NaN                  NaN   
1794  677328882937298944                    NaN                  NaN   

                      timestamp  \
248   2017-03-24 22:08:59 +0000   
849   2016-08-17 20:27:34 +0000   
1926  2015-12-08 02:29:37 +0000   
464   2017-01-06 17:00:38 +0000   
1794  2015-12-17 03:26:04 +0000   

                                                 source  \
248   <a href="http://twitter.com/download/iphone" r...   
849   <a href="http://twitter.com/download/iphone" r...   
1926  <a href="http://twitter.com/download/iphone" r...   
464   <a href="http://twitter.com/download/iphone" r...   
1794  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
248   Say hello to Mimosa. She's an emotional suppor...                  NaN   
849   This is Angel. She stole the @ShopWeRateDogs s...                  NaN   
1926  This is Stanley. Yes he is aware of the spoon'...                  NaN   
464   Meet Strudel. He's rather h*ckin pupset that y...                  NaN   
1794  This is Derek. All the dogs adore Derek. He's ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
248                        NaN                        NaN   
849                        NaN                        NaN   
1926                       NaN                        NaN   
464                        NaN                        NaN   
1794                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
248   https://www.gofundme.com/help-save-a-pup,https...                13   
849   https://twitter.com/dog_rates/status/766008592...                11   
1926  https://twitter.com/dog_rates/status/674053186...                10   
464   https://www.gofundme.com/help-strudel-walk-aga...                11   
1794  https://twitter.com/dog_rates/status/677328882...                10   

      rating_denominator     name  doggo floofer pupper puppo  
248                   10   Mimosa  doggo    None   None  None  
849                   10    Angel   None    None   None  None  
1926                  10  Stanley   None    None   None  None  
464                   10  Strudel   None    None   None  None  
1794                  10    Derek   None    None   None  None

In [12]:
df_twitter_arc.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [13]:
df_twitter_arc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

<a id="twitter-arc:info"></a>
* Dados incompletos, contém apenas 2356 registros dos 5000 anunciados
* `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id` e `retweeted_status_user_id` em `float`
* `timestamp` e `retweeted_status_timestamp` em `string`

In [14]:
# Verificar se há IDs duplicados
df_twitter_arc[df_twitter_arc.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

<a id="twitter-arc:source"></a>

In [15]:
# Visualizar exemplos de dados da coluna `source`
df_twitter_arc.loc[:, 'source'].sample(10)

974     <a href="https://about.twitter.com/products/tw...
2133    <a href="http://twitter.com/download/iphone" r...
1593    <a href="http://twitter.com/download/iphone" r...
1201    <a href="http://vine.co" rel="nofollow">Vine -...
1112    <a href="http://twitter.com/download/iphone" r...
604     <a href="http://twitter.com/download/iphone" r...
1357    <a href="http://twitter.com/download/iphone" r...
1876    <a href="http://twitter.com/download/iphone" r...
1455    <a href="http://twitter.com/download/iphone" r...
1862    <a href="http://twitter.com/download/iphone" r...
Name: source, dtype: object

Variável `source` é uma URL para a Aplicação utilizada no tweet, sem valor para a unidade de observação

<a id="twitter-arc:name"></a>

In [16]:
# Visualizar exemplos da coluna `name`
df_twitter_arc.name.sample(10)

286      Riley
2036      None
386       None
763     Sophie
1311    Tucker
831        Lou
2096       Gin
137       None
1464      None
1083    Bayley
Name: name, dtype: object

In [17]:
df_twitter_arc[df_twitter_arc.name == 'None'].name.count()

745

* Nomes (`name`) preenchidos com valor literal igual a `'None'`, supostamente um valor não preenchido, vazio

<a id="twitter-arc:category"></id>

In [18]:
# Visualizar exemplo de dados das colunas 'doggo' a 'puppo'
df_twitter_arc.loc[:,'doggo':].sample(10)

doggo floofer  pupper  puppo
539    None    None    None   None
1840   None    None    None   None
446   doggo    None    None   None
413    None    None    None  puppo
1847   None    None    None   None
1309   None    None    None   None
1772   None    None  pupper   None
166    None    None    None   None
624   doggo    None    None   None
1306   None    None    None   None

* Valores não preenchidos como 'None' literal

Uma variável em várias colunas

In [19]:
# Verificar padrão no preenchimento das classificações
df_twitter_arc.loc[:,'doggo':].nunique()

doggo      2
floofer    2
pupper     2
puppo      2
dtype: int64

In [20]:
df_twitter_arc.loc[:, 'doggo':'puppo'].isna().sum()

doggo      0
floofer    0
pupper     0
puppo      0
dtype: int64

<a id="twitter-arc:expanded_urls"></a>

Exemplos da coluna `expanded_urls`

In [21]:
df_twitter_arc.loc[:, 'expanded_urls'].sample(20)

1119    https://twitter.com/dog_rates/status/731285275...
445     https://twitter.com/dog_rates/status/819227688...
1292    https://twitter.com/dog_rates/status/708109389...
1656    https://twitter.com/dog_rates/status/683357973...
988     https://twitter.com/dog_rates/status/748977405...
1570                        https://vine.co/v/iOuMphL5DBY
1355    https://twitter.com/dog_rates/status/703611486...
2074    https://twitter.com/dog_rates/status/670842764...
1014    https://twitter.com/dog_rates/status/747204161...
1655    https://twitter.com/dog_rates/status/683391852...
289     https://twitter.com/dog_rates/status/820749716...
1266    https://twitter.com/dog_rates/status/709852847...
804     https://twitter.com/dog_rates/status/772117678...
335     http://us.blastingnews.com/news/2017/02/jfk-an...
820     https://twitter.com/dog_rates/status/770414278...
1900    https://twitter.com/dog_rates/status/674664755...
125     https://twitter.com/dog_rates/status/868622495...
1843    https:

In [22]:
df_twitter_arc.loc[:, 'expanded_urls'].isna().value_counts()

False    2297
True       59
Name: expanded_urls, dtype: int64

* Registros de `expanded_urls` com URLs para os twittes originais. Não apresenta valor a unidade em observacão

------
<a id="avaliacao:image-prediction"></a>

## Avaliação: Image prediction

In [23]:
df_prediction.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

<a id="image-prediction:info"></info>

In [24]:
df_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [25]:
df_prediction.sample(10)

tweet_id                                            jpg_url  \
437   674416750885273600    https://pbs.twimg.com/media/CVwCdCFW4AUHY4D.jpg   
1107  723673163800948736    https://pbs.twimg.com/media/CgsA5eFWgAAu0qn.jpg   
122   668221241640230912    https://pbs.twimg.com/media/CUX_rAyWsAYZOQ5.jpg   
1709  818145370475810820    https://pbs.twimg.com/media/C1qi26rW8AMaj9K.jpg   
2031  883117836046086144    https://pbs.twimg.com/media/DEF2-_hXoAAs62q.jpg   
605   679862121895714818    https://pbs.twimg.com/media/CW9a_h1WwAApmAy.jpg   
2007  877736472329191424    https://pbs.twimg.com/media/DC5YqoQW0AArOLH.jpg   
381   673317986296586240    https://pbs.twimg.com/media/CVgbIobUYAEaeI3.jpg   
827   693486665285931008  https://pbs.twimg.com/ext_tw_video_thumb/69348...   
1455  777621514455814149    https://pbs.twimg.com/media/Csqqoo5WEAAMTVW.jpg   

      img_num                        p1   p1_conf  p1_dog  \
437         1                 Chihuahua  0.287201    True   
1107        1          golden_retriever  0.839390    True   
122         1                      chow  0.395101    True   
1709        1          golden_retriever  0.621931    True   
2031        2          golden_retriever  0.949562    True   
605         1               EntleBucher  0.523206    True   
2007        2  Chesapeake_Bay_retriever  0.837956    True   
381         2        miniature_pinscher  0.384099    True   
827         1                  sea_lion  0.519811   False   
1455        1                      chow  0.999823    True   

                              p2   p2_conf  p2_dog                   p3  \
437                  Boston_bull  0.250920    True              whippet   
1107          Labrador_retriever  0.065706    True          hand_blower   
122             golden_retriever  0.372115    True   Labrador_retriever   
1709          Labrador_retriever  0.364997    True              redbone   
2031          Labrador_retriever  0.045948    True               kuvasz   
605   Greater_Swiss_Mountain_dog  0.431657    True          Appenzeller   
2007          Labrador_retriever  0.062034    True           Weimaraner   
381                   bloodhound  0.079923    True           Rottweiler   
827                  Siamese_cat  0.290971   False  black-footed_ferret   
1455             Norwich_terrier  0.000056    True           Pomeranian   

       p3_conf  p3_dog  
437   0.141012    True  
1107  0.012941   False  
122   0.148785    True  
1709  0.003971    True  
2031  0.002471    True  
605   0.044208    True  
2007  0.040599    True  
381   0.068594    True  
827   0.039967   False  
1455  0.000028    True

<a id="image-prediction:p-values"></a>

Valores não padronizados para as variáveis `p1`, `p2` e `p3`

In [26]:
# Visualização de exemplos de valores nas colunas `p1`, `p2` e `p3`
df_prediction.loc[:, ['p1', 'p2', 'p3']].sample(5)

p1                        p2                      p3
351   golden_retriever                      chow           Irish_terrier
163     shower_curtain  Chesapeake_Bay_retriever                 bathtub
62    golden_retriever          miniature_poodle         standard_poodle
655            snorkel                loggerhead             scuba_diver
1590  English_springer          Brittany_spaniel  Welsh_springer_spaniel

<a id="image-prediction:duplicated-urls"></a>

Verificar a duplicidade de Tweet ID (`tweet_id`) e Imagens (`jpg_url`)

In [27]:
# Verificar a duplicidade de IDs
df_prediction.tweet_id.nunique()

2075

Verificar a duplicidade na coluna `jpg_url`

In [28]:
df_prediction.jpg_url.duplicated().value_counts()

False    2009
True       66
Name: jpg_url, dtype: int64

Há 66 imagens duplicadas

In [29]:
# Validar indicadores de confiabiliade da predição
df_prediction.loc[:, ['p1_conf', 'p2_conf', 'p3_conf']].max()

p1_conf    1.000000
p2_conf    0.488014
p3_conf    0.273419
dtype: float64

------
<a id="avaliacao:twitter-api"></a>

## Avaliação: Twitter API Requests

In [30]:
df_tweets_api.describe()

id  in_reply_to_status_id  in_reply_to_user_id  \
count  1.820000e+03           2.200000e+01         2.200000e+01   
mean   7.237270e+17           6.992047e+17         4.196984e+09   
std    5.777841e+16           4.409222e+16         0.000000e+00   
min    6.660209e+17           6.671522e+17         4.196984e+09   
25%    6.747671e+17           6.747625e+17         4.196984e+09   
50%    7.008223e+17           6.799651e+17         4.196984e+09   
75%    7.617466e+17           7.032024e+17         4.196984e+09   
max    8.924206e+17           8.558181e+17         4.196984e+09   

       retweet_count  favorite_count  
count    1820.000000     1820.000000  
mean     2512.144505     6820.872527  
std      4891.076169    11900.984651  
min        11.000000        0.000000  
25%       536.000000     1363.500000  
50%      1118.000000     3153.000000  
75%      2526.250000     7463.250000  
max     82905.000000   163034.000000

<a id="tweets-api:info"></a>

In [31]:
df_tweets_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1820 entries, 0 to 1819
Data columns (total 10 columns):
id                       1820 non-null int64
created_at               1820 non-null object
in_reply_to_status_id    22 non-null float64
in_reply_to_user_id      22 non-null float64
is_quote_status          1820 non-null bool
retweet_count            1820 non-null int64
favorite_count           1820 non-null int64
media_url                1820 non-null object
retweeted                1820 non-null bool
favorited                1820 non-null bool
dtypes: bool(3), float64(2), int64(3), object(2)
memory usage: 104.9+ KB


In [32]:
df_tweets_api.sample(5)

id                      created_at  \
780   712097430750289920  Tue Mar 22 02:03:52 +0000 2016   
767   714251586676113411  Mon Mar 28 00:43:43 +0000 2016   
179   813910438903693312  Wed Dec 28 00:52:25 +0000 2016   
1432  673352124999274496  Sun Dec 06 04:03:51 +0000 2015   
285   792913359805018113  Mon Oct 31 02:17:31 +0000 2016   

      in_reply_to_status_id  in_reply_to_user_id  is_quote_status  \
780                     NaN                  NaN            False   
767                     NaN                  NaN            False   
179                     NaN                  NaN            False   
1432                    NaN                  NaN            False   
285                     NaN                  NaN            False   

      retweet_count  favorite_count  \
780            1091            3932   
767             880            3385   
179            2034            9842   
1432            548            1669   
285            4370           15246   

                                           media_url  retweeted  favorited  
780   http://pbs.twimg.com/media/CeHg1klW8AE4YOB.jpg      False      False  
767   http://pbs.twimg.com/media/CemIBpfWEAA-TV9.jpg      False      False  
179   http://pbs.twimg.com/media/C0uXObSXUAAIzmV.jpg      False      False  
1432  http://pbs.twimg.com/media/CVg6L2hWIAAYuEb.jpg      False      False  
285   http://pbs.twimg.com/media/CwD-eCFWEAAAJox.jpg      False      False

* `created_at` como `string`
* `in_reply_to_status_id` e `in_reply_to_userid` como `Float`

<a id="tweets-api:duplicated"></a>
Verificar a duplicidade de valores

In [33]:
# Verificar se há IDs duplicados
df_tweets_api.id.duplicated().value_counts()

False    1820
Name: id, dtype: int64

In [34]:
# Verificar se há imagens duplicadas
df_tweets_api.media_url.duplicated().value_counts()

False    1759
True       61
Name: media_url, dtype: int64

Há tweets referenciando a mesma imagem

------

<a id="avaliacao:anotacoes"></a>

## Notas da avaliação
### Qualidade

#### `df_twitter_arc` - Twitter archive
1. [Base incompleta, contém apenas **2356** registros dos **5000** anunciados](#twitter-arc:info)
* [Coluna `source` sem valor à unidade de obsevação](#twitter-arc:source)
* [Nomes (`name`) preenchidos com valor literal igual a `'None'`, vazio](#twitter-arc:name)
* [Coluna `expanded_urls` sem valor à unidade de observação](#twitter-arc:expanded_urls)
* [Categorias \[`doggo`, `floofer`, `pupper`, `puppo`\] com literal `'None'`](#twitter-arc:category)


#### `df_prediction` - Image prediction
1. [Previsões \[`p1`, `p2`, `p3`\] com nomes não padronizados](#image-prediction:p-values)
* [Duplicidade na coluna `jgp_url`](#image-prediction:duplicated-urls)


#### `df_tweets_api` - Twitter API
* [Diferentes Tweets referenciando a mesma imagem, duplicidade](#tweets-api:duplicated)


### Organização

#### `df_twitter_arc` - Twitter archive
1. [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#twitter-arc:info)
* [As colunas `timestamp` e `retweet_status_timestamp` estão em `string`](#twitter-arc:info)
* [Uma variável em várias colunas, `doggo`, `floofer`, `pupper` e `puppo`](#twitter-arc:category)


#### `df_prediction` - Image prediction
1. [Colunas como variáveis na identificação das predições](#image-prediction:info)


#### `df_tweets_api` - Twitter API
1. [Data de criação (`created_at`) como `string`](#tweets-api:info)
* [Valor do id na coluna `in_reply_to_status_id` como `float`](#tweets-api:info)
* [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#tweets-api:info)

------

<a id="limpeza"></a>

# Limpeza

<a id="limpeza:definicao"></a>

## Limpeza: Definição 

#### `df_twitter_arc` - Twitter archive
1. [Base incompleta, de **5000**, apenas **2356** estão disponíveis](#limpeza:twitter-arc:completeness)
    * OK - Nada a ser feito
* [Coluna `source` sem valor à unidade de observação](#limpeza:twitter-arc:source)
    * O conteúdo apresenta um link para download do App Twitter, portanto não agrega valor à unidade de observação, a coluna pode ser removida
* [Nomes (`name`) não preenchidos com valor literal igual a `'None'`](#limpeza:twitter-arc:name)
    * Alterar valores como 'None' para `np.nan`
* [Coluna `expanded_urls` sem valor à unidade de observação](#limpeza:twitter-arc:expanded_urls)(#twitter-arc:expanded_urls)
    * Não apresentam valor à unidade de observação, são URLs para os tweets em questão. Remover a coluna `expanded_urls`.
* [Categorias \[doggo, floofer, pupper, puppo\] com literal `'None'`](#limpeza:twitter-arc:category)
    * Aplicar valor `np.nan` para nomes iguais a 'None'   
* [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#limpeza:twitter-arc:info)
    * A coluna, `retweeted_status_id` ou `retweeted_status_user_id`, podem ser convertidas para `Boolean` como auxílio a identificação de tweets originais
    * Após a criação de uma coluna indicando se o tweet é um retweet, as colunas de retweet, `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`, podem ser descartadas
    * As colunas `in_reply_to_status_id` e `in_reply_to_user_id` serão retratadas no dataframe `df_tweets_api`
* [As colunas `timestamp` e `retweet_status_timestamp` estão em `string` e deveriam ser do tipo `datetime`](#limpeza:twitter-arc:datetime)
    * Converter valores para `datetime`
* [Uma variável em várias colunas, `doggo`, `floofer`, `pupper` e `puppo`](#limpeza:twitter-arc:category-merge)
    * Reduzir a variável para uma coluna, `category`


#### `df_prediction` - Image prediction
1. [Previsões \[`p1`, `p2`, `p3`\] com nomes não padronizados](#limpeza:image-prediction:names)
    * Converter os valores para 'lower case' e substituir '_' por espaços
* [Previsões repetidas sobre a mesma imagem](#limpeza:image-prediction:duplicated-urls)
    * Descartar predições sobre imagens repetidas
* [Colunas como variáveis na identificação das predições](#limpeza:image-prediction:info)
    * Transpor as variáveis nas colunas para linhas


#### `df_tweets_api` - Twitter API
1. [Data de criação (`created_at`) como `string`](#tweets-api:info)
    * Converter datas em string para o formato `datetime`
* [Valor do id na coluna `in_reply_to_status_id` e `in_reply_to_user_id` como `float`](#tweets-api:info)
    * Converter para Inteiro as colunas `in_reply_to_status_id` e `in_reply_to_user_id` e atribuir 0 as que estiverem sem valor


<a id="limpeza:twitter-archive"></a>

## Limpeza: Twitter archive

In [35]:
# Criar cópia do dataframe para a limpeza
df_arch_clean = df_twitter_arc.copy()

<a id="limpeza:twitter-arc:completeness"></a>

### 1. [Base incompleta, de 5000, apenas 2356 estão disponíveis](#limpeza:twitter-arc:completeness)
Nada a ser feito

In [36]:
df_arch_clean.shape

(2356, 17)

<a id="limpeza:twitter-arc:source"></a>

### 2. [Coluna `source` sem valor à unidade de observação](#limpeza:twitter-arc:source)
* O conteúdo apresenta um link para download do App Twitter, portanto não agrega valor à unidade de observação, a coluna pode ser removida

In [37]:
# Remoção da coluna `source`
df_arch_clean.drop('source', axis='columns', inplace=True)

#### Teste: Remoção coluna `source`

In [38]:
# Conferir que a coluna `source` foi removida
df_arch_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

<a id="limpeza:twitter-arc:name"></a>

### 3. [Nomes (`name`) não preenchidos com valor literal igual a `'None'`](#limpeza:twitter-arc:name)
    * Alterar valores como 'None' para `np.nan`

In [39]:
# Contagem de nomes iguais a 'None'
df_arch_clean[df_arch_clean.name.str.lower() == 'none'].name.count()

745

In [40]:
# Função para substituição de valores literais 'None' por np.nan
def set_nan_at_none(value):
    if str(value).lower() == 'none':
        return np.nan
    else:
        return value

In [41]:
# Aplicar `np.nan` aos nomes iguais a 'None'
df_arch_clean.name = df_arch_clean.name.apply(set_nan_at_none)

#### Teste: Aplicação de `np.nan` aos nomes como 'None'

In [42]:
df_arch_clean.name.isnull().sum()

745

In [43]:
df_arch_clean[df_arch_clean.name.str.lower() == 'none'].name.count()

0

In [44]:
df_arch_clean.name.sample(10)

429        NaN
14      Stuart
296       Lucy
969     Stewie
1573      Marq
751        NaN
879       Theo
480      Akumi
1664       NaN
1508       NaN
Name: name, dtype: object

<a id="limpeza:twitter-arc:expanded_urls"></a>

### 4. [Coluna `expanded_urls` sem valor à unidade de observação](#limpeza:twitter-arc:expanded_urls)
* Não apresentam valor à unidade de observação, são URLs para os tweets em questão. Remover a coluna `expanded_urls`

In [45]:
df_arch_clean.drop('expanded_urls', axis='columns', inplace=True)

#### Teste: Remoção da coluna `expanded_urls`

In [46]:
df_arch_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'rating_numerator', 'rating_denominator',
       'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

<a id="limpeza:twitter-arc:category"></a>

### 5. [Categorias \[`doggo`, `floofer`, `pupper`, `puppo`\] com literal `'None'`](#limpeza:twitter-arc:category)
* Aplicar valor `np.nan` para nomes iguais a 'None'   

In [47]:
# Trocar valores 'None' para np.nan nas colunas `doggo`, `floofer`, `pupper`, `puppo`
df_arch_clean.doggo = df_arch_clean.doggo.apply(set_nan_at_none)
df_arch_clean.doggo.isna().value_counts()

True     2259
False      97
Name: doggo, dtype: int64

In [48]:
df_arch_clean.floofer = df_arch_clean.floofer.apply(set_nan_at_none)
df_arch_clean.floofer.isna().value_counts()

True     2346
False      10
Name: floofer, dtype: int64

In [49]:
df_arch_clean.pupper = df_arch_clean.pupper.apply(set_nan_at_none)
df_arch_clean.pupper.isna().value_counts()

True     2099
False     257
Name: pupper, dtype: int64

In [50]:
df_arch_clean.puppo = df_arch_clean.puppo.apply(set_nan_at_none)
df_arch_clean.puppo.isna().value_counts()

True     2326
False      30
Name: puppo, dtype: int64

In [51]:
# Soma de valores preenchidos nas classificações
sum_categories = 0
sum_categories += df_arch_clean.doggo.notna().sum()
sum_categories += df_arch_clean.floofer.notna().sum()
sum_categories += df_arch_clean.pupper.notna().sum()
sum_categories += df_arch_clean.puppo.notna().sum()

sum_categories

394

Apenas 394 dos 2346 registros foram classificados

<a id="limpeza:twitter-arc:info"></a>

### 6. [As colunas `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id` estão em `float64`](#limpeza:twitter-arc:info)
* Linhas com algum valor para as colunas `retweeted_status_id` ou `retweeted_status_user_id` podem ser descartadas por indicarem retweet, e o objetivo é analisar somente tweets originais
* Após a remoção retweets, as colunas `retweeted_status_id`, `retweeted_status_user_id`, `retweeted_status_timestamp`, podem ser descartadas
* As colunas `in_reply_to_status_id` e `in_reply_to_user_id` serão retratadas no dataframe `df_tweets_api`

#### Identificação dos tweets originais
df_arch_clean.loc[:,'retweeted_status_id'].isna().value_counts()

In [52]:
df_arch_clean.retweeted_status_id.size

2356

In [53]:
df_arch_clean.loc[:,'retweeted_status_id'].isna().value_counts()

True     2175
False     181
Name: retweeted_status_id, dtype: int64

Dos 2356 registros, 181 são retweets

In [54]:
# Remoção dos retweets
df_arch_clean.drop(df_arch_clean[df_arch_clean.retweeted_status_id.notna()].index, axis=0, inplace=True)

#### Teste: Remoção de Retweets

In [55]:
# Consultar a existência de colunas com retweet_status_preenchido
df_arch_clean.loc[:,'retweeted_status_id'].isna().value_counts()

True    2175
Name: retweeted_status_id, dtype: int64

Remoção das colunas referentes a retweets

In [56]:
df_arch_clean.drop(['retweeted_status_id', 'retweeted_status_user_id', 'retweeted_status_timestamp'], axis='columns', inplace=True)

#### Teste: Verificação da remoção das colunas referentes a retweets

In [57]:
df_arch_clean.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'text', 'rating_numerator', 'rating_denominator', 'name', 'doggo',
       'floofer', 'pupper', 'puppo'],
      dtype='object')

<a id="limpeza:twitter-arc:datetime"></a>

### 7. [As colunas `timestamp` e `retweet_status_timestamp` estão em `string` e deveriam ser do tipo `datetime`](#limpeza:twitter-arc:datetime)
* Converter valores para `datetime`

Haja vista que a coluna `retweet_status_timestamp` foi removida no passo anterior, não será necessária a sua conversão

In [58]:
df_arch_clean.loc[:, ['timestamp']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 1 columns):
timestamp    2175 non-null object
dtypes: object(1)
memory usage: 34.0+ KB


In [59]:
df_arch_clean.timestamp = pd.to_datetime(df_arch_clean.timestamp)

#### Teste: Veriificação da conversão de `string` para `datetime`

In [60]:
df_arch_clean.loc[:, ['timestamp']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 1 columns):
timestamp    2175 non-null datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1)
memory usage: 34.0 KB


<a id="limpeza:twitter-arc:category-merge"></a>

### 8. [Uma variável em várias colunas, `doggo`, `floofer`, `pupper` e `puppo`](#limpeza:twitter-arc:category-merge)
* Reduzir a variável para uma coluna, `category`

In [61]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null datetime64[ns, UTC]
text                     2175 non-null object
rating_numerator         2175 non-null int64
rating_denominator       2175 non-null int64
name                     1495 non-null object
doggo                    87 non-null object
floofer                  10 non-null object
pupper                   234 non-null object
puppo                    25 non-null object
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(6)
memory usage: 220.9+ KB


In [62]:
df_arch_clean.loc[:, 'doggo':].notna().sum()

doggo       87
floofer     10
pupper     234
puppo       25
dtype: int64

In [63]:
df_arch_clean.loc[:, 'doggo':].notna().sum().sum()

356

De **2175** registros, apenas **356** estão classificados

Criar uma lista das categorias para mesclar as colunas, considerando a coluna a esquerda como prioritaria sobre valores nulos a direita.

In [64]:
category = df_arch_clean.doggo
category = category.combine_first(df_arch_clean.floofer)
category = category.combine_first(df_arch_clean.pupper)
category = category.combine_first(df_arch_clean.puppo)

df_arch_clean['category'] = category

In [65]:
df_arch_clean.category.notna().sum()

344

Após a unificação das colunas de classificação, pode-se ver que resultou em apenas **344** registros, diferente dos **356** iniciais. O que indica que haviam registros com mais de uma classificação.

In [66]:
df_arch_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis='columns', inplace=True)

#### Teste: Verificar a criação da coluna `category` a remoção das colunas `doggo`, `floofer`, `pupper` e `puppo`. 

In [67]:
df_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 9 columns):
tweet_id                 2175 non-null int64
in_reply_to_status_id    78 non-null float64
in_reply_to_user_id      78 non-null float64
timestamp                2175 non-null datetime64[ns, UTC]
text                     2175 non-null object
rating_numerator         2175 non-null int64
rating_denominator       2175 non-null int64
name                     1495 non-null object
category                 344 non-null object
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(3)
memory usage: 169.9+ KB


------
<a id="limpeza:image-prediction"></a>

## Limpeza: Image prediction 

In [68]:
# Cópia do dataframe para limpeza
df_predict_clean = df_prediction.copy()

<a id="limpeza:image-prediction:names"></a>

### 1. [Previsões \[`p1`, `p2`, `p3`\] com nomes não padronizados](#limpeza:image-prediction:names)
* Converter os valores para 'lower case' e substituir '_' por espaços

In [69]:
clean_prediction_name = lambda value: str(value).lower().replace('_', ' ')

In [70]:
df_predict_clean.loc[:, ['p1', 'p2', 'p3']].sample(5)

p1                  p2                p3
182                 cuirass         breastplate  bulletproof_vest
1919  flat-coated_retriever  Labrador_retriever       groenendael
234               Chihuahua                corn            bolete
1889               malamute      Siberian_husky        Eskimo_dog
1752               Pembroke            Cardigan           basenji

Aplicar a padronização dos nomes em cada uma das colunas de predição

In [71]:
df_predict_clean.p1 = df_predict_clean.p1.apply(clean_prediction_name)

In [72]:
df_predict_clean.p2 = df_predict_clean.p2.apply(clean_prediction_name)

In [73]:
df_predict_clean.p3 = df_predict_clean.p3.apply(clean_prediction_name)

#### Teste: Verificar padronização dos nomes das predições

In [74]:
df_predict_clean.loc[:, ['p1', 'p2', 'p3']].sample(5)

p1                              p2                   p3
1629          pembroke             rhodesian ridgeback             cardigan
1028          malamute                      eskimo dog             pembroke
1632    siberian husky  american staffordshire terrier           eskimo dog
979   blenheim spaniel                        shih-tzu  black-footed ferret
759           malinois                 german shepherd               saluki

<a id="limpeza:image-prediction:duplicated-urls"></a>

### 2. [Predições repetidas sobre a mesma imagem](#limpeza:image-prediction:duplicated-urls)
* Descartar predições sobre imagens repetidas

In [75]:
df_predict_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [76]:
df_predict_clean.loc[:, ['jpg_url', 'p1', 'p2', 'p3']].duplicated().value_counts()

False    2009
True       66
dtype: int64

Dos **2075** registros, **66** são sobre a mesma imagem. Entretanto, seguindo o objetivo de trabalhar com apenas Tweets originais, além de descartar duplicidade é interessante também descartar as refêrencias a Retweets. Para isso, será feita uma consulta a seleção de Tweets originais, que foi efetuada no passo [6 da limpeza do Twitter Archive](#limpeza:twitter-arc:info)

In [77]:
df_predict_clean = df_predict_clean.loc[df_predict_clean.tweet_id.isin(df_arch_clean.tweet_id)]

#### Teste: Seleção de Tweets originais

In [78]:
df_predict_clean.tweet_id.isin(df_arch_clean.tweet_id).value_counts()

True    1994
Name: tweet_id, dtype: int64

Apenas dos **2075**, **1994** são de Tweets originais

------

Dos Tweets originai ainda é necessário verificar a existência de duplicidade de predições.

In [79]:
df_predict_clean.loc[:, ['jpg_url']].duplicated().value_counts()

False    1994
dtype: int64

Nenhuma duplicidade foi encontrada após a remoção das referências de Retweets

In [80]:
df_predict_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1994 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    1994 non-null int64
jpg_url     1994 non-null object
img_num     1994 non-null int64
p1          1994 non-null object
p1_conf     1994 non-null float64
p1_dog      1994 non-null bool
p2          1994 non-null object
p2_conf     1994 non-null float64
p2_dog      1994 non-null bool
p3          1994 non-null object
p3_conf     1994 non-null float64
p3_dog      1994 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 161.6+ KB


<a id="limpeza:image-prediction:info"></a>

### 3. [Colunas como variáveis na identificação das predições](#limpeza:image-prediction:info)
* 3.1. Transpor as variáveis nas colunas para linhas:
    * 3.1.1. [`[p1, p2, p3]`, para `prediction_try`](#limpeza:image-prediction:try)
    * 3.1.2. [`[p1_conf, p2_conf, p3_conf]`, para `confidence`](#limpeza:image-prediction:conf)
    * 3.1.3. [`[p1_dog, p2_dog, p3_dog]`, para `is_a_dog`](#limpeza:image-prediction:dog)
* 3.2. [Unir os dataframes de prediction](#limpeza:image-prediction:merge)

In [81]:
df_predict_clean.sample(5)

tweet_id                                            jpg_url  \
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
1474  780459368902959104    https://pbs.twimg.com/media/CtS_p9kXEAE2nh8.jpg   
1111  724046343203856385    https://pbs.twimg.com/media/CgxUTS_XEAAC0pv.jpg   
1046  713175907180089344    https://pbs.twimg.com/media/CeW1tERWAAAA9Q2.jpg   
389   673359818736984064    https://pbs.twimg.com/media/CVhBLohWEAAXtYl.jpg   

      img_num              p1   p1_conf  p1_dog                p2   p2_conf  \
2059        1         whippet  0.626152    True            borzoi  0.194742   
1474        1      great dane  0.382491    True   german shepherd  0.312026   
1111        1           boxer  0.826272    True      bull mastiff  0.158595   
1046        1     timber wolf  0.503788   False          malamute  0.430624   
389         1  english setter  0.696568    True  brittany spaniel  0.104046   

      p2_dog              p3   p3_conf  p3_dog  
2059    True          saluki  0.027351    True  
1474    True    bull mastiff  0.033272    True  
1111    True      great dane  0.011859    True  
1046    True  siberian husky  0.028454    True  
389     True    ibizan hound  0.034832    True

<a id="limpeza:image-prediction:try"></a>

#### 3.1.1. [`[p1, p2, p3]`, para `prediction_try`](#limpeza:image-prediction:try)

In [82]:
# Organizando as variáveis de tentativa de predição e seu palpite
df_predict_guess = pd.melt(df_predict_clean,
        id_vars=['tweet_id', 'jpg_url'],
        value_vars=['p1', 'p2', 'p3'],
        var_name=['prediction_try'],
        value_name='prediction'
       )

In [83]:
df_predict_try.sample(5)

tweet_id                                            jpg_url  \
5319  758467244762497024  https://pbs.twimg.com/ext_tw_video_thumb/75846...   
3499  789599242079838210    https://pbs.twimg.com/media/CvU4UZpXgAE1pAV.jpg   
3221  745712589599014916    https://pbs.twimg.com/media/CllNnkWWMAEDIAR.jpg   
1894  869227993411051520    https://pbs.twimg.com/media/DBAePiVXcAAqHSR.jpg   
2026  666428276349472768    https://pbs.twimg.com/media/CT-g-0DUwAEQdSn.jpg   

     prediction_try                      prediction  
5319             p3  american staffordshire terrier  
3499             p2                          beagle  
3221             p2                            chow  
1894             p1                        pembroke  
2026             p2                            chow

Conversão dos valores da tentativa para número inteiro

In [84]:
# Metodo para retirar o índice da tentativa dos nomes das colunas
get_prediction_index = lambda value: int(re.sub(r'[^\d]', '', str(value)))

In [85]:
df_predict_try.prediction_try = df_predict_try.prediction_try.apply(get_prediction_index)

In [86]:
df_predict_try.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5982 entries, 0 to 5981
Data columns (total 4 columns):
tweet_id          5982 non-null int64
jpg_url           5982 non-null object
prediction_try    5982 non-null int64
prediction        5982 non-null object
dtypes: int64(2), object(2)
memory usage: 187.0+ KB


In [87]:
df_predict_try.sample(5)

tweet_id                                            jpg_url  \
3397  771380798096281600    https://pbs.twimg.com/media/CrR-vVfXEAAk6Gg.jpg   
5791  845397057150107648    https://pbs.twimg.com/media/C7t0IzLWkAINoft.jpg   
4999  709556954897764353    https://pbs.twimg.com/media/CdjaSFCWAAAJZh3.jpg   
813   692752401762250755  https://pbs.twimg.com/tweet_video_thumb/CZ0mhd...   
1181  738883359779196928    https://pbs.twimg.com/media/CkEKe3QWYAAwoDy.jpg   

      prediction_try                   prediction  
3397               2                border collie  
5791               3  west highland white terrier  
4999               3           labrador retriever  
813                1                      samoyed  
1181               1           labrador retriever

<a id="limpeza:image-prediction:conf"></a>

#### 3.1.2. [`[p1_conf, p2_conf, p3_conf]`, para `confidence`](#limpeza:image-prediction:conf)

In [88]:
# Movendo o nível de confiança das predições para uma coluna apenas
df_predict_conf = pd.melt(df_predict_clean,
                         id_vars=['tweet_id', 'jpg_url'],
                         value_vars=['p1_conf', 'p2_conf', 'p3_conf'],
                         var_name=['prediction_try'],
                         value_name='confidence'
                        )

In [89]:
df_predict_conf.sample(5)

tweet_id                                          jpg_url  \
1983  890006608113172480  https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg   
1878  864197398364647424  https://pbs.twimg.com/media/C_4-8iPV0AA1Twg.jpg   
2283  671182547775299584  https://pbs.twimg.com/media/CVCE9uYXIAEtSzR.jpg   
3169  737826014890496000  https://pbs.twimg.com/media/Cj1I1fbWYAAOwff.jpg   
1596  808106460588765185  https://pbs.twimg.com/media/Czb4iFRXgAIUMiN.jpg   

     prediction_try  confidence  
1983        p1_conf    0.957979  
1878        p1_conf    0.945905  
2283        p2_conf    0.218601  
3169        p2_conf    0.005605  
1596        p1_conf    0.426183

In [90]:
# Atribuir apenas valores inteiros aos indicadores de tentativas
df_predict_conf.prediction_try = df_predict_conf.prediction_try.apply(get_prediction_index)

In [91]:
df_predict_conf.sample(5)

tweet_id                                          jpg_url  \
4377  673576835670777856  https://pbs.twimg.com/media/CVkGjsxU8AA5OYX.jpg   
2996  708845821941387268  https://pbs.twimg.com/media/CdZTgynWwAATZcx.jpg   
5625  815390420867969024  https://pbs.twimg.com/media/C1DZQiTXgAUqgRI.jpg   
3018  710833117892898816  https://pbs.twimg.com/media/Cd1i8qvUkAE-Jlr.jpg   
5596  810984652412424192  https://pbs.twimg.com/media/C0EyPZbXAAAceSc.jpg   

      prediction_try  confidence  
4377               3    0.072735  
2996               2    0.167853  
5625               3    0.046147  
3018               2    0.189712  
5596               3    0.025823

In [92]:
df_predict_conf.shape

(5982, 4)

<a id="limpeza:image-prediction:dog"></a>

#### 3.1.3. [`[p1_dog, p2_dog, p3_dog]`, para `is_a_dog`](#limpeza:image-prediction:dog)

In [93]:
# Movendo o resultado da predição, `p#_dog` para `is_a_dog`
df_predict_dog = pd.melt(df_predict_clean,
                        id_vars=['tweet_id', 'jpg_url'],
                        value_vars=['p1_dog', 'p2_dog', 'p3_dog'],
                        var_name=['prediction_try'],
                        value_name='is_a_dog')

In [94]:
df_predict_dog.sample(5)

tweet_id                                          jpg_url  \
423   674075285688614912  https://pbs.twimg.com/media/CVrL5YBWoAA_uPD.jpg   
2816  693231807727280129  https://pbs.twimg.com/media/CZ7aplIUsAAq-8s.jpg   
1687  823581115634085888  https://pbs.twimg.com/media/C23ypm6VQAAO31l.jpg   
3264  750026558547456000  https://pbs.twimg.com/media/CmieRQRXgAA8MV3.jpg   
1429  775842724423557120  https://pbs.twimg.com/media/CsRY1jAWYAUOx55.jpg   

     prediction_try  is_a_dog  
423          p1_dog      True  
2816         p2_dog      True  
1687         p1_dog     False  
3264         p2_dog     False  
1429         p1_dog      True

In [95]:
# Atribuir apenas valores inteiros aos indicadores de tentativas
df_predict_dog.prediction_try = df_predict_dog.prediction_try.apply(get_prediction_index)

In [96]:
df_predict_dog.sample(5)

tweet_id                                          jpg_url  \
5617  813217897535406080  https://pbs.twimg.com/media/C0khWkVXEAI389B.jpg   
1731  831309418084069378  https://pbs.twimg.com/media/C4lngK5VUAEVrNO.jpg   
4302  671743150407421952  https://pbs.twimg.com/media/CVKC1IfWIAAsQks.jpg   
2628  681242418453299201  https://pbs.twimg.com/media/CXRCXesVAAArSXt.jpg   
4018  666418789513326592  https://pbs.twimg.com/media/CT-YWb7U8AA7QnN.jpg   

      prediction_try  is_a_dog  
5617               3      True  
1731               1      True  
4302               3     False  
2628               2     False  
4018               3      True

In [97]:
df_predict_dog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5982 entries, 0 to 5981
Data columns (total 4 columns):
tweet_id          5982 non-null int64
jpg_url           5982 non-null object
prediction_try    5982 non-null int64
is_a_dog          5982 non-null bool
dtypes: bool(1), int64(2), object(1)
memory usage: 146.1+ KB


<a id="limpeza:image-prediction:merge"></a>

#### 3.2. [Unir os dataframes de prediction](#limpeza:image-prediction:merge)

In [100]:
# Lista de dataframes para merge
predictions = [
    df_predict_try,
    df_predict_conf,
    df_predict_dog
]

In [ ]:
# Merge
df_predict_clean = pd.merge(prediction)

------
<a id="limpeza:twitter-api"></a>

## Limpeza: Twitter API

------
<a id="armazenamento"></a>

# Armazenamento

# Relatórios

* Data wrangling efforts
* Analyses and visualizations